### ISSUE TITLE
 I don't know what the exact Train Step is.

### ISSUE THREAD

**Issue by wjddyd66** [tongli1210/BreastCancerSurvivalIntegration#2]
In Paper... (Integrating the complementary information: concatenation autoencoder (ConcatAE))

>For each modality, we train an independent autoencoder and transform the input features into a hidden space. We then concatenate the hidden features from each modality and feed the concatenated hidden feature into the task-specific network. Compared to the single-modality network, we have a separate reconstruction loss for each data modality

But, i see on your code. (OS_Autoencoder_Concat.py)

```python
def _train_on_batch(self, epoch, batch_idx, batch):
        net_q, net_p, net_c = self.net_q, self.net_p, self.net_c
        opt_q, opt_p, opt_c = self.optimizer_q, self.optimizer_p, self.optimizer_c
        for i in range(self.n_modality):
            net_q[i].train()
            net_p[i].train()
        net_c.train()   

        X_list = [None]*self.n_modality
        X_recon_list = [None]*self.n_modality
        z_list = [None]*self.n_modality
        loss_mse_list = [None]*self.n_modality
        # read in training data
        for i in range(self.n_modality):
            X_list[i] = batch[self.modality[i]]
        X_list = [tmp.cuda() for tmp in X_list]
        survival_time = batch['days'].cuda()
        survival_event = batch['event'].cuda()
        batch_size = X_list[0].shape[0]
        batch_log = {'size': batch_size}
       
        #################################################### 
        # Stage 1: train Q and P with reconstruction loss  #
        ####################################################
        for i in range(self.n_modality):
            net_q[i].zero_grad()
            net_p[i].zero_grad()
            for p in net_q[i].parameters():
                p.requires_grad=True
            for p in net_p[i].parameters():
                p.requires_grad=True
        for p in net_c.parameters():
            p.requires_grad=False

        t0 = time()
        for i in range(self.n_modality):
            z_list[i] = net_q[i](X_list[i])
            X_recon_list[i] = net_p[i](z_list[i])
            loss_mse = nn.functional.mse_loss(X_recon_list[i], X_list[i]) # Mean square error
            loss_mse_list[i] = loss_mse.item()
            loss_mse.backward()
            opt_q[i].step()
            opt_p[i].step()
        t_mse = time() - t0
        batch_log['loss_mse'] = sum(loss_mse_list)

        ####################################################
        # Stage 2: train Q and C with classification loss  #
        ####################################################
        if not survival_event.sum(0): # skip the batch if all instances are negative
            batch_log['loss_survival'] = torch.Tensor([float('nan')])
            return batch_log 
        for i in range(self.n_modality):
            net_q[i].zero_grad()
            for p in net_q[i].parameters():
                p.requires_grad=True
            for p in net_p[i].parameters():
                p.requires_grad=False
        net_c.zero_grad()
        for p in net_c.parameters():
            p.requires_grad=True

        t0 = time()
        for i in range(self.n_modality):
            z_list[i] = net_q[i](X_list[i])
        z_combined = torch.cat(z_list, dim=1)
        pred = net_c(z_combined)

        loss_survival = neg_par_log_likelihood(pred, survival_time, survival_event)
        loss_survival.backward()
        for i in range(self.n_modality):
            opt_q[i].step()
        opt_c.step() 
        t_survival = time() - t0
        c_index = CIndex(pred, survival_event, survival_time)
        batch_log['loss_survival'] = loss_survival.item()
        batch_log['c_index'] = c_index

        if self.opt.log_time:
            batch_log['t_mse'] = t_mse
            batch_log['t_survival'] = t_survival
        return batch_log
```

**I am wondering how the exact Train-Step is achieved.**
1. Finish AutoEncoder learning. -> Training the Classification Model.
2. Train AutoEncoder and Classification Model together.

I think it is number 2 in the code....
